cfy_manager/components/cluster.py

```diff
@@ -12,25 +12,159 @@
+                request.get_method = lambda: 'PUT'
+                response = urllib2.urlopen(request)
+            else:
+                response = None
+        # keep an erroneous HTTP response to examine its status code, but still
+        # abort on fatal errors like being unable to connect at all
+        except urllib2.HTTPError as e:
+            response = e
+        except urllib2.URLError as e:
+            raise NetworkError(
+                'REST service returned an invalid response: {0}'.format(e))
+        if response.code == 401:
+            raise NetworkError(
+                'Could not connect to the REST service: '
+                '401 unauthorized. Possible access control misconfiguration'
+            )
+        if response.code != 200:
+            raise NetworkError(
+                'REST service returned an unexpected response: '
+                '{0}'.format(response.code)
+            )
+
+        try:
+            return json.load(response)
+        except ValueError as e:
+            raise BootstrapError(
+                'REST service returned malformed JSON: {0}'.format(e))
+
+    def _verify_premium(self):
+        try:
+            find_module('cloudify_premium')
+        except ImportError as e:
+            premium_enabled = False
+        else:
+            premium_enabled = True
+        return premium_enabled
+
+    def _get_join_addresses(self, master_manager_ip):
+        result = self._generic_cloudify_rest_request(
+            master_manager_ip,
+            config[RESTSERVICE]['port'],
+            'v3.1/cluster/nodes',
+            'get')
+        return [n.host_ip for n in result['items']]
+
+    def _adding_new_cluster_node_on_master(self, master_manager_ip,
+                                           slave_host_ip):
+        data = {
+            'host_ip': master_manager_ip,
+            'node_name': slave_host_ip
+        }
+        result = self._generic_cloudify_rest_request(
+            master_manager_ip,
+            config[RESTSERVICE]['port'],
+            'v3.1/cluster/nodes/{0}'.format(slave_host_ip),
+            'put',
+            data
+        )
+        return result
+
+    def _get_cluster_status(self, master_manager_ip):
+        result = self._generic_cloudify_rest_request(
+            master_manager_ip,
+            config[RESTSERVICE]['port'],
+            'v3.1/cluster',
+            'get')
+        return result
+
+    def _join_to_cluster(self, master_manager_ip):
+        cluster_node_host_ip = config[MANAGER][PUBLIC_IP]
+        cluster_node_name = config[CLUSTER][NODE_NAME]
+        join_addrs = self._get_join_addresses(master_manager_ip)
+        new_node = self._adding_new_cluster_node_on_master(master_manager_ip,
+                                                           join_addrs)
+        data = {
+            'host_ip': cluster_node_host_ip,
+            'node_name': cluster_node_name,
+            'credentials': new_node['credentials'],
```

Comment:
`'credentials'` and afair also `'required'` (requirements ie. only allow nodes with riemann to join a master that has riemann)

Reply:
fixed